File: java/src/main/java/problems201_300/ImplementQueueUsingStacks.java
package problems201_300;

import java.util.Stack;

class ImplementQueueUsingStacks {
    private final Stack<Integer> pushStack = new Stack<>();
    private final Stack<Integer> popStack = new Stack<>();

    public void push(int x) {
        pushStack.push(x);
    }

    public int pop() {
        fillPopStack();
        return popStack.pop();
    }

    public int peek() {
        fillPopStack();
        return popStack.peek();
    }

    public boolean empty() {
        return pushStack.isEmpty() && popStack.isEmpty();
    }

    private void fillPopStack() {
        if (!popStack.isEmpty()) {
            return;
        }
        while (!pushStack.isEmpty()) {
            popStack.push(pushStack.pop());
        }
    }
}

File: kotlin/src/main/kotlin/problems201_300/ImplementQueueUsingStacks.kt
package problems201_300

import java.util.Stack

class ImplementQueueUsingStacks {
    private val pushStack = Stack<Int>()
    private val popStack = Stack<Int>()

    fun push(x: Int) {
        pushStack.push(x)
    }

    fun pop(): Int {
        fillPopStack()
        return popStack.pop()
    }

    fun peek(): Int {
        fillPopStack()
        return popStack.peek()
    }

    fun empty(): Boolean = pushStack.isEmpty() && popStack.isEmpty()

    private fun fillPopStack() {
        if (!popStack.isEmpty()) {
            return
        }
        while (!pushStack.isEmpty()) {
            popStack.push(pushStack.pop())
        }
    }
}
